File: javascript/leetcode/27. Remove Element.js
/**
 * @param {number[]} nums
 * @param {number} val
 * @return {number}
 */
var removeElement = function(nums, val) {
  let first = 0;
  let last = nums.length - 1;
  let len = 0;

  while (first <= last) {
    if (nums[last] === val) {
      nums[last--] = val - 1;
      len++;
    } else {
      if (nums[first] === val) {
        nums[first++] = nums[last];
        nums[last--] = val - 1;
        len++;
      } else first++;
    }
  }

  return nums.length - len;
};

File: c/leetcode/27. Remove Element.c


int removeElement(int *nums, int numsSize, int val)
{
  int first = 0;
  int last = numsSize - 1;
  int len = 0;

  while (first <= last)
  {
    if (nums[last] == val)
    {
      nums[last--] = val - 1;
      len++;
    }
    else
    {
      if (nums[first] == val)
      {
        nums[first++] = nums[last];
        nums[last--] = val - 1;
        len++;
      }
      else
        first++;
    }
  }

  return numsSize - len;
}
